src/backend/distributed/transaction/multi_shard_transaction.c

```diff
@@ -43,18 +43,95 @@ void
 OpenTransactionsToAllShardPlacements(List *shardIntervalList, char *userName)
 {
 	ListCell *shardIntervalCell = NULL;
+	List *newConnectionList = NIL;
+	ListCell *connectionCell = NULL;
 
 	if (shardConnectionHash == NULL)
 	{
 		shardConnectionHash = CreateShardConnectionHash(TopTransactionContext);
 	}
 
+	BeginOrContinueCoordinatedTransaction();
+	if (MultiShardCommitProtocol == COMMIT_PROTOCOL_2PC)
+	{
+		CoordinatedTransactionUse2PC();
+	}
+
+	/* open connections to shards which don't have connections yet */
 	foreach(shardIntervalCell, shardIntervalList)
 	{
 		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
 		uint64 shardId = shardInterval->shardId;
+		ShardConnections *shardConnections = NULL;
+		bool shardConnectionsFound = false;
+		List *shardPlacementList = NIL;
+		ListCell *placementCell = NULL;
+
+		shardConnections = GetShardConnections(shardId, &shardConnectionsFound);
+		if (shardConnectionsFound)
+		{
+			continue;
+		}
+
+		shardPlacementList = FinalizedShardPlacementList(shardId);
+		if (shardPlacementList == NIL)
+		{
+			/* going to have to have some placements to do any work */
+			ereport(ERROR, (errmsg("could not find any shard placements for the shard "
+								   UINT64_FORMAT, shardId)));
+		}
 
-		BeginTransactionOnShardPlacements(shardId, userName);
+		foreach(placementCell, shardPlacementList)
+		{
+			ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(placementCell);
+			MultiConnection *connection = NULL;
+			MemoryContext oldContext = NULL;
+
+			WorkerNode *workerNode = FindWorkerNode(shardPlacement->nodeName,
+													shardPlacement->nodePort);
+			if (workerNode == NULL)
+			{
+				ereport(ERROR, (errmsg("could not find worker node %s:%d",
+									   shardPlacement->nodeName,
+									   shardPlacement->nodePort)));
+			}
+
+			connection = StartNodeUserDatabaseConnection(FORCE_NEW_CONNECTION,
+														 shardPlacement->nodeName,
+														 shardPlacement->nodePort,
+														 userName,
+														 NULL);
+
+			/* we need to preserve the connection list for the next statement */
```

Comment:
Weird that you changed this comment from _whole top-level transaction_ to _next statement_. Is that meaningful?